src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -31,53 +30,61 @@ public CompositeHandler(PostgresCompositeType postgresType, ConnectorTypeMapper
             _nameTranslator = nameTranslator;
         }
 
-        public override async ValueTask<T> Read(NpgsqlReadBuffer buffer, int length, bool async, FieldDescription? fieldDescription = null)
+        public override ValueTask<T> Read(NpgsqlReadBuffer buffer, int length, bool async, FieldDescription? fieldDescription = null)
         {
-            if (_memberHandlers == null)
-                _memberHandlers = CreateMemberHandlers();
-
-            await buffer.Ensure(sizeof(int), async);
+            Initialize();
 
-            var fieldCount = buffer.ReadInt32();
-            if (fieldCount != _memberHandlers.Length)
-                throw new InvalidOperationException($"pg_attributes contains {_memberHandlers.Length} fields for type {PgDisplayName}, but {fieldCount} fields were received.");
+            return _constructorHandler is null
```

Comment:
Alternative:

```suggestion
            return _constructorHandler?.Read(buffer, async) ?? ReadUsingMemberHandlers();
```

Reply:
Tried, but it doesn't make the code better. Let's leave as is for now. I would like to allow to mix materialization via the constructor and property binding, so this code should go away.